woocommerce/woocommerce-ios
# Core Data Stack updates

Closes: #14054 
Also closes: #2370 


## Description

This PR adds a few improvements to the Core Data stack: 
- Updated the `writerDerivedStorage` to be independent of the main context. Instead, the background context is now directly associated with the persistent container following Apple's [recommendation](https://href.li/?https://developer.apple.com/documentation/coredata/using_core_data_in_the_background#2904035). The main context now merges changes from the persistent container automatically.
- A writer operation queue has been introduced to take care of write operations one-by-one following WP/JP's implementation [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/d30b801b5be1f23db0f2143a27389f0966f27a7e/WordPress/Classes/Utility/ContextManager.swift#L118). I'm following the simple variation for the completion handler without the result parameter as we don't currently have any failure handling for the write operations - I suppose we can add that variation when we need it.
- Removed the unused method `performBackgroundTask` - we want to avoid creating too many background context through this method.
- Updated `saveDerivedType` to only save the background context to the persistent container.
- Marked access to `writerDerivedStorage` and `saveDerivedType` as deprecated and added a helper method `performAndSave` for performing write operations similarly to WP/JP's Core Data stack. Updates for the stores will be handled in future PRs.
- Added check to ensure write operations are handled in a background context. The checks are handled in debug mode only, and a new launch argument `-enforce-core-data-write-in-background` has been added to enable `assert` when writing is done on the main context.
- Initialize the storage manager along with `AppDelegate` to ensure that the Core Data stack is created before the stores manager. This is an attempt to eliminate confusing stack traces [peaMlT-Rx-p2#comment-2194]. The order of initialization should now be as follows: CrashLogging > CoreDataManager > DefaultStoresManager.
- Keep product's `date` and `dateCreated` Swift properties optional to avoid crashes when transforming deleted product objects in Core Data.

TODO: @itsmeichigo to update the release notes before merging.

## Steps to reproduce

Smoke testing the app
- With `-enforce-core-data-write-in-background` disabled, build and run the app in debug mode. 
- Log in to a test store with existing products/orders/Blaze campaigns.
- Smoke test the app following the guide P91TBi-bVe-p2.

Testing enforcing write in background
- With `-enforce-core-data-write-in-background` enabled, build and run the app in debug mode. 
- Log in to a test store with existing products/orders/Blaze campaigns.
- Confirm that the `assert` causes a fatal error where a write operation is handled in the main context.

## Testing information

Tested on device iPad pro 4th gen iOS 17.6:
- Logged in to an atomic store with wpcom account successfully.
- Confirmed that dashboard works as before: stats cards with different time ranges, all cards enabled with different filters.
- Confirmed that orders work normally: pagination, searching, order creation and edit, creating shipping labels, marking orders as complete.
- Confirmed that products work normally: pagination, searching, bulk update, product creation with AI, creating & editing variable products, updating product details.
- Confirmed that hub menu works normally: store switching, settings, payments, google ads, blaze, woo admin, view store, coupons, reviews, inbox, customers.

Tested on simulator iPhone 15 Pro iOS 17.4 and `-enforce-core-data-write-in-background` enabled and confirmed that write operations in the main context causes assertion failure.

## Screenshots

N/A

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:
